torch/_inductor/fx_passes/pad_mm.py

```diff
@@ -318,32 +398,68 @@ def should_pad_mm(match: Match) -> bool:
     )
 
 
-def mm_replace(mat1: Tensor, mat2: Tensor) -> Tensor:
-    m_padded_length = get_padded_length(mat1.shape[0], get_alignment_size(mat1))
-    k_padded_length = get_padded_length(mat1.shape[1], get_alignment_size(mat1))
-    n_padded_length = get_padded_length(mat2.shape[1], get_alignment_size(mat2))
-
-    return pad_mm(mat1, mat2, m_padded_length, k_padded_length, n_padded_length)
-
-
 def pad_mm(
     mat1: Tensor,
     mat2: Tensor,
     m_padded_length: int,
     k_padded_length: int,
     n_padded_length: int,
+    explicit_transpose: bool = False,
 ) -> Tensor:
-    # mm_replace will go through pad_mm multiple times if multiple dimensions are needed to be padded
-    if k_padded_length != 0:
-        mat1 = pad_dim(mat1, k_padded_length, 1)
-        mat2 = pad_dim(mat2, k_padded_length, 0)
-        return torch.ops.aten.mm(mat1, mat2)
-    elif n_padded_length != 0:
-        mat2 = pad_dim(mat2, n_padded_length, 1)
-        return torch.ops.aten.mm(mat1, mat2)[:, :-n_padded_length]
+    if k_padded_length != 0 or m_padded_length != 0:
+        # dim order is reversed for constant_pad_nd, for every dim we specify right and left padding
+        mat1_padded = aten.constant_pad_nd(
+            mat1, [0, k_padded_length, 0, m_padded_length]
+        )
+    else:
+        mat1_padded = mat1
+    if k_padded_length != 0 or n_padded_length != 0:
+        # dim order is reversed for constant_pad_nd, for every dim we specify right and left padding
+        mat2_padded = aten.constant_pad_nd(
+            mat2, [0, n_padded_length, 0, k_padded_length]
+        )
     else:
-        mat1 = pad_dim(mat1, m_padded_length, 0)
-        return torch.ops.aten.mm(mat1, mat2)[:-m_padded_length, :]
+        mat2_padded = mat2
+    if explicit_transpose:
+        # If M dimension is aligned but N is not, this is an alternative to a padding N
+        # which has the advantage of enabling downstream epilogue fusions
+        # padding on K dim, transpose and contiguous should be fuseable into a single op
+        res = torch.ops.aten.mm(
+            mat2_padded.transpose(-1, -2).contiguous(), mat1_padded.transpose(-1, -2)
+        ).transpose(-1, -2)
+    else:
+        res = torch.ops.aten.mm(mat1_padded, mat2_padded)
+    if m_padded_length != 0:
+        res = res[:-m_padded_length, :]
+    if n_padded_length != 0:
+        res = res[:, :-n_padded_length]
+    return res
+
+
+def mm_replace(mat1: Tensor, mat2: Tensor) -> Tensor:
+    k_padded_length = get_padded_length(mat1.shape[1], get_alignment_size(mat1))
+    explicit_transpose = False
+    if not torch._inductor.config.shape_pad_only_k_dim:
```

Comment:
if we dont have any results of padding only k dim (maybe the opbench inputs will have one: https://github.com/pytorch/pytorch/blob/main/benchmarks/dynamo/microbenchmarks/operator_inp_logs/hf_train/BigBird_training.txt) do you think it's worth removing this config option ?

Reply:
removed it